platform/mellanox/onie-fw-update

```diff
@@ -0,0 +1,150 @@
+#!/bin/sh
+
+# Copyright (C) 2019 Mellanox Technologies Ltd.
+# Copyright (C) 2019 Michael Shych <michaelsh@mellanox.com>
+#
+# SPDX-License-Identifier:     GPL-2.0
+
+this_script=${ONIE_FWPKG_PROGRAM_NAME:-$(basename $(realpath $0))}
+
+onie_mount=/mnt/onie-boot
+os_boot=/host
+onie_partition=
+onie_entry=0
+
+export ONIE_FWPKG_PROGRAM_NAME=$(basename $(realpath $0))
+
+usage()
+{
+cat <<EOF
+  update
+    The 'update' command will reboot system to ONIE update mode
+    and ONIE will perform automatically update of previously 
+    added (i.e. pending) FW (ONIE itself, BIOS or CPLD) image.
+
+EOF
+}
+
+enable_onie_access()
+{
+	onie_partition=$(fdisk -l | grep "ONIE boot" | awk '{print $1}')
+	if [ ! -d $onie_mount ]; then
+		mkdir /mnt/onie-boot
+	fi
+	mount $onie_partition /mnt/onie-boot
+	if [ ! -e /lib/onie ]; then
+		ln -s /mnt/onie-boot/onie/tools/lib/onie /lib/onie
+	fi
+	PATH=/sbin:/usr/sbin:/bin:/usr/bin:$onie_mount/onie/tools/bin/
+	export PATH
+}
+
+clean_onie_access()
+{
+	rm -f /lib/onie
+	umount $onie_partition
+}
+
+# ONIE entry must exist in grub config
+find_onie_menuentry()
+{
+	onie_entry="$(cat $os_boot/grub/grub.cfg | grep -e 'menuentry' | cat -n | awk '$0~/ONIE/ {print $1-1}')"
+	entries_num="$(echo "$onie_entry" | grep -E '^[0-9]+$' | wc -l)"
+	if [ $entries_num -eq 1 ] && [ $onie_entry -ge 1 ]; then
+		return 0
+	fi
+	return 1
+}
+
+change_grub_boot_order()
+{
+	find_onie_menuentry
+	rc=$?
+	if [ $rc -eq 0 ]; then
+		grub-reboot --boot-directory=$os_boot $onie_entry
+	else
+		echo "ERROR: ONIE entry wasn't found in grub config"
+		return 1
+	fi
+
+	grub-editenv $onie_mount/grub/grubenv set onie_mode=update
+	return 0
+}
+
+show_pending()
+{
+	curr_dir=$(pwd)
+	cd $onie_mount/onie/update/pending || return 0
+	num=$(find . -type f | wc -l)
+	if [ $num -ge 1 ]; then
+		echo "Number of FW update pending files are: "$num
+		ls -l * | awk {'print $9"  "$5"  "$7"-"$6" "$8'}
+	else
+		echo "There is no pending files for FW update."
+	fi
+	cd $curr_dir
+
+	return $num
+}
+
+system_reboot()
+{
+	echo "Reboot will be done after 5 sec."
+	sleep 5
+	/sbin/reboot
+}
+
+# Process command arguments
+cmd=$1
+# optional argument
+name="$2"
+
+if [ -z "$cmd" ] ; then
+	# Default to 'show' if no command is specified.
+	cmd="show"
+fi
+
+case "$cmd" in
+	add | remove)
+		[ -z "$name" ] && {
+			echo "ERROR: This command requires a firmware update file name."
+			echo "Run '$this_script help' for complete details."
+			exit 1
+		}
+		;;
+	update)
+		enable_onie_access
+		show_pending
+		rc=$?
+		if [ $rc -ne 0 ]; then
+			change_grub_boot_order
+			rc=$?
+			clean_onie_access
+			if [ $rc -eq 0 ]; then
+				system_reboot
```

Comment:
@stephenxs You need to update the script, because it will do the reboot regardless the user action.

Reply:
ok. will remove it.

Reply:
done.

Reply:
@nazariig just FYI, I just remember that after a BIOS upgrade scheduled, the boot sequence of grub will be updated to 2 which will not be recognized by sonic_installer. As a result, any attempt trying to reboot switch via a SONiC command, like "sudo reboot", won't work. The only way to reboot the system is to use /sbin/reboot. 
This hint has been printed to the screen after a BIOS upgrade successfully scheduled.